src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs

```diff
@@ -44,40 +44,80 @@ public class InternalBufferSizeTests : FileSystemWatcherTest
+                int internalBufferOperationCapacity = CalculateInternalBufferOperationCapacity(watcher.InternalBufferSize, file.Path);
 
                 // Set the capacity high to ensure no error events arise.
                 if (setToHigherCapacity)
                     watcher.InternalBufferSize = watcher.InternalBufferSize * 12;
 
-                // block the handling thread
-                ManualResetEvent unblockHandler = new ManualResetEvent(false);
-                watcher.Changed += (o, e) =>
-                {
-                    unblockHandler.WaitOne();
-                };
-
-                // generate enough file change events to overflow the default buffer
-                Action action = () =>
-                {
-                    for (int i = 1; i < internalBufferOperationCapacity * 10; i++)
-                    {
-                        File.SetLastWriteTime(file.Path, DateTime.Now + TimeSpan.FromSeconds(i));
-                    }
-
-                    unblockHandler.Set();
-                };
-                Action cleanup = () => unblockHandler.Reset();
+                Action action = GetAction(unblockHandler, internalBufferOperationCapacity, file.Path);
+                Action cleanup = GetCleanup(unblockHandler);
 
                 if (setToHigherCapacity)
                     ExpectNoError(watcher, action, cleanup);
                 else
                     ExpectError(watcher, action, cleanup);
             }
         }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]  // Uses P/Invokes
+        public void FileSystemWatcher_InternalBufferSize_SynchronizingObject()
+        {
+            TestISynchronizeInvoke invoker = new TestISynchronizeInvoke();
+            ManualResetEvent unblockHandler = new ManualResetEvent(false);
+            using (var testDirectory = new TempDirectory(GetTestFilePath()))
+            using (var file = new TempFile(Path.Combine(testDirectory.Path, "file")))
+            using (FileSystemWatcher watcher = CreateWatcher(testDirectory.Path, file.Path, unblockHandler))
+            {
+                watcher.SynchronizingObject = invoker;
+
+                int internalBufferOperationCapacity = CalculateInternalBufferOperationCapacity(watcher.InternalBufferSize, file.Path);
+
+                Action action = GetAction(unblockHandler, internalBufferOperationCapacity, file.Path);
+                Action cleanup = GetCleanup(unblockHandler);
+
+                ExpectError(watcher, action, cleanup);
+                Assert.True(invoker.BeginInvoke_Called);
+            }
+        }
+
+        #region Test Helpers
+
+        private FileSystemWatcher CreateWatcher(string testDirectoryPath, string filePath, ManualResetEvent unblockHandler)
+        {
+            var watcher = new FileSystemWatcher(testDirectoryPath, Path.GetFileName(filePath));
+
+            // block the handling thread
+            watcher.Changed += (o, e) => unblockHandler.WaitOne();
```

Comment:
Will this work with a recreated watcher?

Reply:
I don't understand the question.  Can you elaborate?

Reply:
https://github.com/dotnet/runtime/blob/f044460da04f0833990325b144f8f26478293cca/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L387-L425

https://github.com/dotnet/runtime/blob/f044460da04f0833990325b144f8f26478293cca/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L511-L528

If `watcher` is created in `RecreateWatcher()`, will the correct thread be blocked?

Reply:
Ah, no.  The new watcher you're creating won't have the event handler that was originally stored into the first watcher.

Reply:
I assume I should leave this as is, since this is a pre-existing flaw with `FileSystemWatcher_InternalBufferSize()`?

Reply:
Yeah, we can clean up stuff later if it's an issue.

Reply:
Except, is this actually pre-existing?  Where is the existing FileSystemWatcher_InternalBufferSize recreating the watcher and re-setting this handler?

Reply:
In the `main` branch, `FileSystemWatcher_InternalBufferSize()` calls one of the test methods:

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs#L76-L79

Both of them call `TryErrorEvent()`:

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L351-L355

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L364-L368

`TryErrorEvent()` recreates the watcher, but the handler is never re-set:

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L370-L431